src/Rules/Comparison/ImpossibleCheckTypeHelper.php

```diff
@@ -80,8 +80,38 @@ public function findSpecifiedType(
 					'trait_exists',
 					'enum_exists',
 				], true)) {
+					if (count($node->getArgs()) < 1) {
+						return null;
+					}
+					$argType = $scope->getType($node->getArgs()[0]->value);
+					if (count($argType->getConstantStrings()) !== 1) {
+						return null;
+					}
+					$argConstantString = $argType->getConstantStrings()[0];
+					if (!$this->reflectionProvider->hasClass($argConstantString->getValue())) {
+						return null;
+					}
+					$reflection = $this->reflectionProvider->getClass($argConstantString->getValue());
+
+					if ($functionName === 'class_exists' && ($reflection->isInterface() || $reflection->isTrait())) {
+						return false;
+					}
+
+					if ($functionName === 'interface_exists' && ($reflection->isClass() || $reflection->isTrait() || $reflection->isEnum())) {
+						return false;
+					}
+
+					if ($functionName === 'trait_exists' && ($reflection->isClass() || $reflection->isInterface() || $reflection->isEnum())) {
+						return false;
+					}
+
+					if ($functionName === 'enum_exists' && ($reflection->isClass() || $reflection->isInterface() || $reflection->isTrait())) {
+						return false;
```

Comment:
I think these conditions could be improved. Name what it cannot be instead of what it can be, for example:

```php
if ($functionName === 'interface_exists' && !$reflection->isInterface()) {
```

Reply:
But that can break if some future PHP version onboards new struct type. But that would probably lead to many changes everywhere, so why not...